src/rgw/rgw_sal.h

```diff
@@ -269,18 +404,32 @@ class RGWObject {
     const std::string &get_name() const { return key.name; }
 
     virtual int get_obj_state(RGWObjectCtx *rctx, RGWBucket& bucket, RGWObjState **state, optional_yield y, bool follow_olh = false) = 0;
-    virtual int get_obj_attrs(RGWObjectCtx *rctx, optional_yield y, rgw_obj *target_obj = nullptr) = 0;
+    virtual int set_attrs(RGWObjectCtx* rctx, RGWAttrs* setattrs, RGWAttrs* delattrs, optional_yield y, rgw_obj* target_obj = NULL) = 0;
+    virtual int get_obj_attrs(RGWObjectCtx *rctx, optional_yield y, rgw_obj* target_obj = NULL) = 0;
     virtual int modify_obj_attrs(RGWObjectCtx *rctx, const char *attr_name, bufferlist& attr_val, optional_yield y) = 0;
     virtual int delete_obj_attrs(RGWObjectCtx *rctx, const char *attr_name, optional_yield y) = 0;
     virtual int copy_obj_data(RGWObjectCtx& rctx, RGWBucket* dest_bucket, RGWObject* dest_obj, uint16_t olh_epoch, std::string* petag, const DoutPrefixProvider *dpp, optional_yield y) = 0;
+    virtual bool is_expired() = 0;
 
+    RGWAttrs& get_attrs(void) { return attrs; }
```

Comment:
After get_obj_attrs(), ReadOp::prepare(), or copy_obj_data().  Basically after get_obj_attrs() or any operation that needs attrs.

Reply:
can we find a way to prevent this from being misused?

in general, i'm uncomfortable about these API objects acting as a cache also. in small classes, it's easy to encapsulate mutable state and maintain the invariants. but as classes get really big (like Object), it gets increasingly difficult to reason about all of its state across all of its operations

if an RGWOp needs attributes, it's probably better for it to read them via `get_obj_attrs()` and cache them itself if necessary

Reply:
Okay, I'll look into this for future work.